src/js/view/controls/controls.ts

```diff
@@ -25,24 +28,158 @@ const ACTIVE_TIMEOUT = OS.mobile ? 4000 : 2000;
 const ALWAYS_ALLOWED_KEYS = [27];
 
 ErrorContainer.cloneIcon = cloneIcon;
-instances.forEach(api => {
+instances.forEach((api: PlayerAPI) => {
     if (api.getState() === STATE_ERROR) {
         const errorIconContainer = api.getContainer().querySelector('.jw-error-msg .jw-icon');
-        if (errorIconContainer && !errorIconContainer.hasChildNodes()) {
+        if (errorIconContainer && !errorIconContainer.hasChildNodes() && ErrorContainer.cloneIcon) {
             errorIconContainer.appendChild(ErrorContainer.cloneIcon('error'));
         }
     }
 });
 
-const reasonInteraction = function () {
+const reasonInteraction = () => {
     return { reason: 'interaction' };
 };
-export default class Controls extends Events {
-    constructor(context, playerContainer) {
-        super();
 
-        // Alphabetic order
-        // Any property on the prototype should be initialized here first
+interface BaseControls {
```

Comment:
Can rename.
I added `BaseControls` as an abstract class and moved most of the methods from `Controls` to it. I found when working on the `TizenControls`, that the main methods that were different were `enable` and `disable`, so I made them abstract methods.
I tried to keep the common methods and properties in `BaseControls`, which also extends `Events`, so other controls could inherit from that.

Reply:
do we need an interface if the methods and attrs are defined by the abstract class?